src/main/java/com/spartronics4915/lib/lidar/LidarProcessor.java

```diff
@@ -254,29 +259,35 @@ private void processLidarScan(LidarScan scan)
     {
         try
         {
-            Pose2d p;
+            Pose2d lastPose = mLidarStateMap.getLatestFieldToVehicle().getValue();
+            Pose2d p = null;
             if(mMode == OperatingMode.kRelative)
             {
                 Transform xform = mRelativeICP.doRelativeICP(scan.getPoints());
                 if(xform != null)
                 {
                     p = xform.inverse().toPose2d();
+                    p = p.transformBy(lastPose);
                     Logger.debug("relativeICP: " + p.toString());
                 }
+                else Logger.warning("Relative ICP returned a null transform!");
             } 
             else
             {
-                IRobotStateMap.State estimate = mStateMap.get(scan.getTimestamp());
+                Pose2d estimate = mEncoderStateMap.getFieldToVehicle(scan.getTimestamp());
                 Transform xform = mICP.doICP(getCulledPoints(scan), 
-                                new Transform(estimate.position).inverse(), 
+                                new Transform(estimate).inverse(), 
                                 mReferenceModel);
                 p  = xform.inverse().toPose2d();
                 Logger.debug("absoluteICP: " + p.toString());
             }
-            // until robot is actually moving, we don't want
-            // to update robot pose. That is, we expect the
-            // "same" point cloud each iteration.
-            // sPoses.put(new InterpolatingDouble(ts), zeroPose);
+
+            if (p != null)
+            {
+                mLidarStateMap.addObservations(scan.getTimestamp(), p,
+                    new Twist2d(p.distance(lastPose), 0, p.getRotation().distance(lastPose.getRotation())),
+                    Twist2d.identity() /* No good way to get predicted velocity */);
```

Comment:
@dbadb Does this seem correct to you? This is dependent on how long ago we sampled `lastPose` (Now that I'm thinking about it, we should probably sample exactly 1 second ago. If you agree, then I'll go ahead and change it). Also, regarding the `dy` of 0, `Kinematics` seems to do it roughly this way (it averages the two wheel velocities and puts those velocities into the `dx` of a `Twist2d`), but this seems like a bit of an abuse of `Twist2d`. What are your thoughts?

Reply:
E.g. I think the correct implementation might be:
```java
                Pose2d pose1SecBeforeScan = mLidarStateMap.getFieldToVehicle(scan.getTimestamp() - 1000);
                mLidarStateMap.addObservations(scan.getTimestamp(), p,
                    new Twist2d(p.distance(pose1SecBeforeScan), 0, p.getRotation().distance(pose1SecBeforeScan.getRotation())),
                    Twist2d.identity() /* No good way to get predicted velocity */);
```

Reply:
definitely worthy of a chat